File: BirthdayBot/Configuration.vb
Imports System.Reflection
Imports Newtonsoft.Json.Linq
Imports System.IO

''' <summary>
''' Loads and holds configuration values.
''' </summary>
Class Configuration
    Public ReadOnly Property BotToken As String
    Public ReadOnly Property LogWebhook As String
    Public ReadOnly Property DBotsToken As String
    Public ReadOnly Property DatabaseSettings As Database

    Public ReadOnly Property ShardCount As Integer

    Sub New()
        ' Looks for settings.json in the executable directory.
        Dim confPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)
        confPath += Path.DirectorySeparatorChar + "settings.json"

        If Not File.Exists(confPath) Then
            Throw New Exception("Settings file not found. " _
                                + "Create a file in the executable directory named 'settings.json'.")
        End If

        Dim jc = JObject.Parse(File.ReadAllText(confPath))

        BotToken = jc("BotToken")?.Value(Of String)()
        If String.IsNullOrWhiteSpace(BotToken) Then
            Throw New Exception("'BotToken' must be specified.")
        End If

        LogWebhook = jc("LogWebhook")?.Value(Of String)()
        If String.IsNullOrWhiteSpace(LogWebhook) Then
            Throw New Exception("'LogWebhook' must be specified.")
        End If

        Dim dbj = jc("DBotsToken")
        If dbj IsNot Nothing Then
            DBotsToken = dbj.Value(Of String)()
        Else
            DBotsToken = Nothing
        End If

        Dim sqlcs = jc("SqlConnectionString")?.Value(Of String)()
        If String.IsNullOrWhiteSpace(sqlcs) Then
            Throw New Exception("'SqlConnectionString' must be specified.")
        End If
        DatabaseSettings = New Database(sqlcs)

        Dim sc? = jc("ShardCount")?.Value(Of Integer)()
        If Not sc.HasValue Then
            ShardCount = 1
        Else
            ShardCount = sc.Value
            If ShardCount <= 0 Then
                Throw New Exception("'ShardCount' must be a positive integer.")
            End If
        End If
    End Sub
End Class

File: Configuration.cs
using CommandLine;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Diagnostics.CodeAnalysis;
using System.Reflection;
using System.Text.RegularExpressions;

namespace BirthdayBot;
/// <summary>
/// Loads and holds configuration values.
/// </summary>
partial class Configuration {
    [GeneratedRegex(@"(?<low>\d{1,2})[-,](?<high>\d{1,2})")]
    private static partial Regex ShardRangeParser();
    const string KeyShardRange = "ShardRange";

    public string BotToken { get; }
    public string? DBotsToken { get; }

    public int ShardStart { get; }
    public int ShardAmount { get; }
    public int ShardTotal { get; }

    public string? SqlHost { get; }
    public string? SqlDatabase { get; }
    public string SqlUsername { get; }
    public string SqlPassword { get; }
    internal string SqlApplicationName { get; }

    /// <summary>
    /// Number of seconds between each time the status task runs, in seconds.
    /// </summary>
    public int StatusInterval { get; }
    /// <summary>
    /// Number of concurrent shard startups to happen on each check.
    /// This value also determines the maximum amount of concurrent background database operations.
    /// </summary>
    public int MaxConcurrentOperations { get; }
    /// <summary>
    /// Amount of time to wait between background task runs within each shard.
    /// </summary>
    public int BackgroundInterval { get; }
    /// <summary>
    /// Gets whether to show common connect/disconnect events and other related messages.
    /// This is disabled in the public instance, but it's worth keeping enabled in self-hosted bots.
    /// </summary>
    public bool LogConnectionStatus { get; }

    public Configuration() {
        var args = CommandLineParameters.Parse(Environment.GetCommandLineArgs());
        var path = args?.ConfigFile ?? Path.GetDirectoryName(Assembly.GetEntryAssembly()!.Location)
            + Path.DirectorySeparatorChar + "." + Path.DirectorySeparatorChar + "settings.json";

        // Looks for configuration file
        JObject jc;
        try {
            var conftxt = File.ReadAllText(path);
            jc = JObject.Parse(conftxt);
        } catch (Exception ex) {
            string pfx;
            if (ex is JsonException) pfx = "Unable to parse configuration: ";
            else pfx = "Unable to access configuration: ";

            throw new Exception(pfx + ex.Message, ex);
        }

        BotToken = ReadConfKey<string>(jc, nameof(BotToken), true);
        DBotsToken = ReadConfKey<string>(jc, nameof(DBotsToken), false);

        ShardTotal = args.ShardTotal ?? ReadConfKey<int?>(jc, nameof(ShardTotal), false) ?? 1;
        if (ShardTotal < 1) throw new Exception($"'{nameof(ShardTotal)}' must be a positive integer.");

        var shardRangeInput = args.ShardRange ?? ReadConfKey<string>(jc, KeyShardRange, false);
        if (!string.IsNullOrWhiteSpace(shardRangeInput)) {
            var m = ShardRangeParser().Match(shardRangeInput);
            if (m.Success) {
                ShardStart = int.Parse(m.Groups["low"].Value);
                var high = int.Parse(m.Groups["high"].Value);
                ShardAmount = high - (ShardStart - 1);
            } else {
                throw new Exception($"Shard range not properly formatted in '{KeyShardRange}'.");
            }
        } else {
            // Default: this instance handles all shards
            ShardStart = 0;
            ShardAmount = ShardTotal;
        }

        SqlHost = ReadConfKey<string>(jc, nameof(SqlHost), false);
        SqlDatabase = ReadConfKey<string?>(jc, nameof(SqlDatabase), false);
        SqlUsername = ReadConfKey<string>(jc, nameof(SqlUsername), true);
        SqlPassword = ReadConfKey<string>(jc, nameof(SqlPassword), true);
        SqlApplicationName = $"Shard{ShardStart:00}-{ShardStart + ShardAmount - 1:00}";

        StatusInterval = ReadConfKey<int?>(jc, nameof(StatusInterval), false) ?? 90;
        MaxConcurrentOperations = ReadConfKey<int?>(jc, nameof(MaxConcurrentOperations), false) ?? 4;
        BackgroundInterval = ReadConfKey<int?>(jc, nameof(BackgroundInterval), false) ?? 60;
        LogConnectionStatus = ReadConfKey<bool?>(jc, nameof(LogConnectionStatus), false) ?? true;
    }

    private static T? ReadConfKey<T>(JObject jc, string key, [DoesNotReturnIf(true)] bool failOnEmpty) {
        if (jc.ContainsKey(key)) return jc[key]!.Value<T>();
        if (failOnEmpty) throw new Exception($"'{key}' must be specified.");
        return default;
    }

    class CommandLineParameters {
        [Option('c', "config")]
        public string? ConfigFile { get; set; }

        [Option("shardtotal")]
        public int? ShardTotal { get; set; }

        [Option("shardrange")]
        public string? ShardRange { get; set; }

        public static CommandLineParameters? Parse(string[] args) {
            CommandLineParameters? result = null;

            new Parser(settings => {
                settings.IgnoreUnknownArguments = true;
                settings.AutoHelp = false;
                settings.AutoVersion = false;
            }).ParseArguments<CommandLineParameters>(args)
                .WithParsed(p => result = p)
                .WithNotParsed(e => { /* ignore */ });
            return result;
        }
    }
}
